JsonB provider not found when loaded from library

**Describe the bug**
I'm using `smallrye-reactive-messaging-provider` library as implementation of the reactive messaging api. The library is packaged into the war file, because the implementation provided by liberty don't provide the new features I require. On a Incoming message channel I want to convert the input stream using Jsonb to an object.
```java
@Incoming("results")
public CompletionStage<Void> getMessages(Message<InputStream> message) {
	JsonbBuilder.create();
}
```
But the Jsonb provider is not found.
In other parts of the code the Jsonb is used without problems.

```
io.smallrye.reactive.messaging.ProcessingException: Exception thrown when calling the method de.uni_stuttgart.tik.viplab.websocket_api.ecs.ECSComputationService#getMessages
	at io.smallrye.reactive.messaging.AbstractMediator.lambda$initialize$0(AbstractMediator.java:50)
	at io.smallrye.reactive.messaging.AbstractMediator.invoke(AbstractMediator.java:61)
	at io.smallrye.reactive.messaging.SubscriberMediator.lambda$processMethodReturningACompletionStage$6(SubscriberMediator.java:161)
	at io.smallrye.reactive.streams.stages.FlatMapCompletionStageFactory$FlatMapCompletionStage.lambda$apply$0(FlatMapCompletionStageFactory.java:47)
	at io.reactivex.internal.operators.flowable.FlowableFlatMap$MergeSubscriber.onNext(FlowableFlatMap.java:132)
	at io.reactivex.internal.operators.flowable.FlowableFlatMap$MergeSubscriber.drainLoop(FlowableFlatMap.java:501)
	at io.reactivex.internal.operators.flowable.FlowableFlatMap$MergeSubscriber.drain(FlowableFlatMap.java:366)
	at io.reactivex.internal.operators.flowable.FlowableFlatMap$InnerSubscriber.onNext(FlowableFlatMap.java:665)
	at io.reactivex.internal.subscriptions.DeferredScalarSubscription.complete(DeferredScalarSubscription.java:117)
	at io.reactivex.processors.AsyncProcessor.subscribeActual(AsyncProcessor.java:242)
	at io.reactivex.Flowable.subscribe(Flowable.java:14918)
	at io.reactivex.Flowable.subscribe(Flowable.java:14865)
	at io.reactivex.internal.operators.flowable.FlowableFlatMap$MergeSubscriber.onNext(FlowableFlatMap.java:163)
	at io.smallrye.reactive.streams.utils.ConnectableProcessor.onNext(ConnectableProcessor.java:122)
	at org.eclipse.microprofile.reactive.streams.operators.CompletionSubscriber$1DefaultCompletionSubscriber.onNext(CompletionSubscriber.java:85)
	at org.eclipse.microprofile.reactive.streams.operators.CompletionSubscriber$1DefaultCompletionSubscriber.onNext(CompletionSubscriber.java:85)
	at io.smallrye.reactive.messaging.SubscriberMediator$1.onNext(SubscriberMediator.java:100)
	at io.smallrye.reactive.streams.utils.WrappedSubscriber.onNext(WrappedSubscriber.java:42)
	at io.reactivex.subscribers.SafeSubscriber.onNext(SafeSubscriber.java:90)
	at io.reactivex.internal.operators.flowable.FlowableFlatMap$MergeSubscriber.drainLoop(FlowableFlatMap.java:501)
	at io.reactivex.internal.operators.flowable.FlowableFlatMap$MergeSubscriber.drain(FlowableFlatMap.java:366)
	at io.reactivex.internal.operators.flowable.FlowableFlatMap$InnerSubscriber.onNext(FlowableFlatMap.java:665)
	at io.reactivex.internal.subscriptions.DeferredScalarSubscription.complete(DeferredScalarSubscription.java:117)
	at io.reactivex.processors.AsyncProcessor.onComplete(AsyncProcessor.java:201)
	at io.smallrye.reactive.streams.utils.CompletionStageToPublisher.lambda$fromCompletionStage$0(CompletionStageToPublisher.java:23)
	at java.base/java.util.concurrent.CompletableFuture.uniWhenComplete(CompletableFuture.java:859)
	at java.base/java.util.concurrent.CompletableFuture$UniWhenComplete.tryFire(CompletableFuture.java:837)
	at java.base/java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:506)
	at java.base/java.util.concurrent.CompletableFuture.complete(CompletableFuture.java:2144)
	at de.uni_stuttgart.tik.ecs.ecc.connector.ECSInput.executePollActions(ECSInput.java:81)
	at de.uni_stuttgart.tik.ecs.ecc.connector.ECSInput.lambda$2(ECSInput.java:57)
	at com.ibm.ws.threading.internal.PolicyTaskFutureImpl.run(PolicyTaskFutureImpl.java:715)
	at com.ibm.ws.threading.internal.PolicyExecutorImpl.runTask(PolicyExecutorImpl.java:1135)
	at com.ibm.ws.threading.internal.PolicyExecutorImpl$GlobalPoolTask.run(PolicyExecutorImpl.java:189)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:830)
Caused by: java.lang.reflect.InvocationTargetException
	at jdk.internal.reflect.GeneratedMethodAccessor671.invoke(Unknown Source)
	at java.base/java.lang.reflect.Method.invoke(Method.java:567)
	at io.smallrye.reactive.messaging.AbstractMediator.lambda$initialize$0(AbstractMediator.java:48)
	... 36 more
Caused by: java.util.ServiceConfigurationError: javax.json.bind.spi.JsonbProvider: Provider org.eclipse.yasson.JsonBindingProvider not found
	at java.base/java.util.ServiceLoader.fail(ServiceLoader.java:590)
	at java.base/java.util.ServiceLoader$LazyClassPathLookupIterator.nextProviderClass(ServiceLoader.java:1212)
	at java.base/java.util.ServiceLoader$LazyClassPathLookupIterator.hasNextService(ServiceLoader.java:1221)
	at java.base/java.util.ServiceLoader$LazyClassPathLookupIterator.hasNext(ServiceLoader.java:1265)
	at java.base/java.util.ServiceLoader$2.hasNext(ServiceLoader.java:1300)
	at java.base/java.util.ServiceLoader$3.hasNext(ServiceLoader.java:1385)
	at javax.json.bind.spi.JsonbProvider.provider(JsonbProvider.java:112)
	at javax.json.bind.JsonbBuilder.create(JsonbBuilder.java:108)
	at de.uni_stuttgart.tik.viplab.websocket_api.ecs.ECSComputationService.getMessages(ECSComputationService.java:69)
	at de.uni_stuttgart.tik.viplab.websocket_api.ecs.ECSComputationService$Proxy$_$$_WeldClientProxy.getMessages(Unknown Source)
	... 39 more
```

**Steps to Reproduce**
maven dependencies
```
	<dependencies>
		<!-- Workaround for Beta features -->
		<dependency>
			<groupId>io.smallrye.reactive</groupId>
			<artifactId>smallrye-reactive-messaging-provider</artifactId>
			<version>1.0.8</version>
		</dependency>
		<dependency>
			<groupId>io.smallrye.reactive</groupId>
			<artifactId>smallrye-reactive-streams-operators</artifactId>
			<version>1.0.10</version>
		</dependency>

		<dependency>
			<groupId>javax.websocket</groupId>
			<artifactId>javax.websocket-api</artifactId>
			<version>1.1</version>
			<scope>provided</scope>
		</dependency>
		<dependency>
			<groupId>javax.json.bind</groupId>
			<artifactId>javax.json.bind-api</artifactId>
			<version>1.0</version>
			<scope>provided</scope>
		</dependency>
		<dependency>
			<groupId>org.slf4j</groupId>
			<artifactId>slf4j-api</artifactId>
			<version>1.7.28</version>
		</dependency>
		<dependency>
			<groupId>javax.enterprise</groupId>
			<artifactId>cdi-api</artifactId>
			<version>2.0</version>
			<scope>provided</scope>
		</dependency>
		<dependency>
			<groupId>javax.annotation</groupId>
			<artifactId>javax.annotation-api</artifactId>
			<version>1.3.2</version>
			<scope>provided</scope>
		</dependency>
		<dependency>
			<groupId>org.eclipse.microprofile.config</groupId>
			<artifactId>microprofile-config-api</artifactId>
			<version>1.3</version>
			<scope>provided</scope>
		</dependency>
		<dependency>
			<groupId>org.eclipse.microprofile.metrics</groupId>
			<artifactId>microprofile-metrics-api</artifactId>
			<version>2.0</version>
			<scope>provided</scope>
		</dependency>
		<!-- <dependency> <groupId>org.eclipse.microprofile.reactive.messaging</groupId> 
			<artifactId>microprofile-reactive-messaging-api</artifactId> <version>1.0</version> 
			<scope>compile</scope> </dependency> -->
		<dependency>
			<groupId>org.eclipse.microprofile.rest.client</groupId>
			<artifactId>microprofile-rest-client-api</artifactId>
			<version>1.3.4</version>
			<scope>compile</scope>
		</dependency>
		<dependency>
			<groupId>javax.ws.rs</groupId>
			<artifactId>javax.ws.rs-api</artifactId>
			<version>2.1.1</version>
			<scope>provided</scope>
		</dependency>

	</dependencies>
```
2. send a message to the incoming channel.
3. Code for handling the message is executed:
```
@Incoming("results")
public CompletionStage<Void> getMessages(Message<InputStream> message) {
	JsonbBuilder.create();
}
```
4. exception is thrown

**Expected behavior**
The Jsonb Provider is found and can be used to convert json strings to objects.

**Diagnostic information:**
 - OpenLiberty Version: 19.0.0.11 and 19.0.0.12
 - Java Version: 8 and 13 tested
```
java version "13.0.1" 2019-10-15
Java(TM) SE Runtime Environment (build 13.0.1+9)
Java HotSpot(TM) 64-Bit Server VM (build 13.0.1+9, mixed mode, sharing)
```
 - server.xml configuration 
```
<server description="name">
	<featureManager>
		<feature>websocket-1.1</feature>
		<feature>jsonb-1.0</feature>
		<feature>cdi-2.0</feature>
		<feature>mpConfig-1.3</feature>
		<feature>mpMetrics-2.0</feature>
		<feature>mpHealth-2.0</feature>
		<feature>concurrent-1.0</feature>
		<!-- <feature>mpReactiveMessaging-1.0</feature> -->
		<!-- <feature>mpRestClient-1.3</feature> -->
	</featureManager>

	<mpMetrics authentication="false"/>
	<httpEndpoint httpPort="${default.http.port}"
		httpsPort="-1" id="defaultHttpEndpoint" onError="FAIL" />
</server>
```
**Additional context**
It seams like a class loading problem, because the context Classloader can't find the SPI provider class `org.eclipse.yasson.JsonBindingProvider`, from this method invocation. But when in another context the SPI provider class `org.eclipse.yasson.JsonBindingProvider` can be loaded without problems.

--- a/dev/com.ibm.ws.cdi.weld/src/com/ibm/ws/cdi/liberty/CDIRuntimeImpl.java
+++ b/dev/com.ibm.ws.cdi.weld/src/com/ibm/ws/cdi/liberty/CDIRuntimeImpl.java
@@ -1,5 +1,5 @@
 /*******************************************************************************
- * Copyright (c) 2015, 2016 IBM Corporation and others.
+ * Copyright (c) 2015, 2020 IBM Corporation and others.
  * All rights reserved. This program and the accompanying materials
  * are made available under the terms of the Eclipse Public License v1.0
  * which accompanies this distribution, and is available at
@@ -13,7 +13,9 @@ package com.ibm.ws.cdi.liberty;
 import java.util.ArrayList;
 import java.util.Iterator;
 import java.util.List;
+import java.util.Map;
 import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
 import java.util.concurrent.ExecutorService;
 import java.util.concurrent.ScheduledExecutorService;
 
@@ -106,6 +108,7 @@ public class CDIRuntimeImpl extends AbstractCDIRuntime implements ApplicationSta
     private boolean isClientProcess;
     private RuntimeFactory runtimeFactory;
     private ProxyServicesImpl proxyServices;
+    private final Map<String, ClassLoader> appTccls = new ConcurrentHashMap<>();
 
     public void activate(ComponentContext cc) {
         containerConfigRef.activate(cc);
@@ -421,9 +424,11 @@ public class CDIRuntimeImpl extends AbstractCDIRuntime implements ApplicationSta
                 this.runtimeFactory.removeApplication(appInfo);
                 return;
             }
-            newCL = getRealAppClassLoader(application);
 
-            if (newCL != null) {
+            ClassLoader appCL = getRealAppClassLoader(application);
+            if (appCL != null) {
+                newCL = classLoadingSRRef.getServiceWithException().createThreadContextClassLoader(appCL);
+                appTccls.put(appInfo.getName(), newCL);
                 oldCl = CDIUtils.getAndSetLoader(newCL);
             }
 
@@ -486,6 +491,14 @@ public class CDIRuntimeImpl extends AbstractCDIRuntime implements ApplicationSta
                 }
             } catch (CDIException e) {
                 //FFDC and carry on
+            } finally {
+                // Clean up the application TCCL created for startup
+                // Must do this at shutdown since it's possible for the app to hold onto it and use it after startup
+                ClassLoader tccl = appTccls.get(appInfo.getName());
+                if (tccl != null) {
+                    classLoadingSRRef.getServiceWithException().destroyThreadContextClassLoader(tccl);
+                    appTccls.remove(appInfo.getName());
+                }
             }
         }
     }

--- a/dev/com.ibm.ws.microprofile.metrics.cdi.2.0/src/io/astefanutti/metrics/cdi20/MetricsExtension.java
+++ b/dev/com.ibm.ws.microprofile.metrics.cdi.2.0/src/io/astefanutti/metrics/cdi20/MetricsExtension.java
@@ -26,8 +26,6 @@ package io.astefanutti.metrics.cdi20;
 import java.lang.annotation.Annotation;
 import java.lang.reflect.Member;
 import java.lang.reflect.Type;
-import java.security.AccessController;
-import java.security.PrivilegedAction;
 import java.util.Collections;
 import java.util.HashMap;
 import java.util.HashSet;
@@ -64,10 +62,6 @@ import org.eclipse.microprofile.metrics.annotation.Counted;
 import org.eclipse.microprofile.metrics.annotation.Gauge;
 import org.eclipse.microprofile.metrics.annotation.Metered;
 import org.eclipse.microprofile.metrics.annotation.Timed;
-import org.osgi.framework.Bundle;
-import org.osgi.framework.BundleContext;
-import org.osgi.framework.FrameworkUtil;
-import org.osgi.framework.ServiceReference;
 import org.osgi.service.component.annotations.Component;
 import org.osgi.service.component.annotations.Reference;
 
@@ -77,7 +71,6 @@ import com.ibm.ws.microprofile.metrics.cdi.decorator.AnnotatedTypeDecorator;
 import com.ibm.ws.microprofile.metrics.cdi20.helper.Utils;
 import com.ibm.ws.microprofile.metrics.cdi20.producer.MetricRegistryFactory;
 import com.ibm.ws.microprofile.metrics.impl.SharedMetricRegistries;
-import com.ibm.wsspi.classloading.ClassLoadingService;
 
 @Component(service = WebSphereCDIExtension.class, immediate = true)
 public class MetricsExtension implements Extension, WebSphereCDIExtension {
@@ -150,27 +143,8 @@ public class MetricsExtension implements Extension, WebSphereCDIExtension {
             Metadata metadata = name.metadataOf(bean.getValue());
             String[] tags = name.tagOf(bean.getValue());
 
-            ClassLoader origLoader = Thread.currentThread().getContextClassLoader();
+            registry.register(metadata, (Metric) getReference(manager, bean.getValue().getBaseType(), bean.getKey()), Utils.tagsToTags(tags)); // line 190
 
-            final Bundle bundle = FrameworkUtil.getBundle(ClassLoadingService.class);
-            ClassLoadingService thingthing = AccessController.doPrivileged(new PrivilegedAction<ClassLoadingService>() {
-                @Override
-                public ClassLoadingService run() {
-                    BundleContext bCtx = bundle.getBundleContext();
-                    ServiceReference<ClassLoadingService> svcRef = bCtx.getServiceReference(ClassLoadingService.class);
-                    return svcRef == null ? null : bCtx.getService(svcRef);
-                }
-            });
-
-            try {
-
-                ClassLoader tccl = thingthing.createThreadContextClassLoader(origLoader);
-                Thread.currentThread().setContextClassLoader(tccl);
-
-                registry.register(metadata, (Metric) getReference(manager, bean.getValue().getBaseType(), bean.getKey()), Utils.tagsToTags(tags)); // line 190
-            } finally {
-                Thread.currentThread().setContextClassLoader(origLoader);
-            }
             MetricID mid = new MetricID(metadata.getName(), Utils.tagsToTags(tags));
             addMetricID(mid);
         }

--- a/dev/com.ibm.ws.microprofile.reactive.messaging_fat/bnd.bnd
+++ b/dev/com.ibm.ws.microprofile.reactive.messaging_fat/bnd.bnd
@@ -25,7 +25,9 @@ fat.project: true
 	com.ibm.websphere.org.eclipse.microprofile.reactive.messaging.1.0;version=latest,\
 	com.ibm.websphere.org.eclipse.microprofile.reactive.streams.operators.1.0;version=latest,\
 	com.ibm.websphere.org.eclipse.microprofile.config.1.3;version=latest,\
+	com.ibm.websphere.javaee.concurrent.1.0;version=latest,\
 	com.ibm.websphere.javaee.cdi.2.0;version=latest,\
+	com.ibm.websphere.javaee.jsonb.1.0;version=latest,\
 	com.ibm.ws.org.slf4j.api.1.7.7;version=latest,\
 	org.apache.kafka:kafka-clients;version=2.3.0,\
 	org.testcontainers:testcontainers;version=1.12.3,\

--- a/dev/com.ibm.ws.microprofile.reactive.messaging_fat/fat/src/com/ibm/ws/microprofile/reactive/messaging/fat/jsonb/JsonbBean.java
+++ b/dev/com.ibm.ws.microprofile.reactive.messaging_fat/fat/src/com/ibm/ws/microprofile/reactive/messaging/fat/jsonb/JsonbBean.java
@@ -1,0 +1,47 @@
+/*******************************************************************************
+ * Copyright (c) 2019 IBM Corporation and others.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ *
+ * Contributors:
+ *     IBM Corporation - initial API and implementation
+ *******************************************************************************/
+package com.ibm.ws.microprofile.reactive.messaging.fat.jsonb;
+
+import javax.enterprise.context.ApplicationScoped;
+import javax.json.bind.Jsonb;
+import javax.json.bind.JsonbBuilder;
+
+import org.eclipse.microprofile.reactive.messaging.Incoming;
+import org.eclipse.microprofile.reactive.messaging.Outgoing;
+import org.eclipse.microprofile.reactive.streams.operators.ProcessorBuilder;
+import org.eclipse.microprofile.reactive.streams.operators.ReactiveStreams;
+
+/**
+ * Bean which uses JsonB to convert objects to json
+ * <p>
+ * Note that the Jsonb object will be created at application startup
+ */
+@ApplicationScoped
+public class JsonbBean {
+
+    @Incoming("input")
+    @Outgoing("step1")
+    public ProcessorBuilder<TestData, String> toJson() {
+        // Tests creating Jsonb at startup
+        Jsonb jsonb = JsonbBuilder.create();
+        return ReactiveStreams.<TestData> builder()
+                        .map((t) -> jsonb.toJson(t));
+    }
+
+    @Incoming("step1")
+    @Outgoing("output")
+    public TestData fromJson(String json) {
+        // Tests creating Jsonb at processing time (app fully started)
+        Jsonb jsonb = JsonbBuilder.create();
+        return jsonb.fromJson(json, TestData.class);
+    }
+
+}

--- a/dev/com.ibm.ws.microprofile.reactive.messaging_fat/fat/src/com/ibm/ws/microprofile/reactive/messaging/fat/jsonb/JsonbDeliveryBean.java
+++ b/dev/com.ibm.ws.microprofile.reactive.messaging_fat/fat/src/com/ibm/ws/microprofile/reactive/messaging/fat/jsonb/JsonbDeliveryBean.java
@@ -1,0 +1,34 @@
+/*******************************************************************************
+ * Copyright (c) 2019 IBM Corporation and others.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ *
+ * Contributors:
+ *     IBM Corporation - initial API and implementation
+ *******************************************************************************/
+package com.ibm.ws.microprofile.reactive.messaging.fat.jsonb;
+
+import java.util.concurrent.CompletionStage;
+
+import javax.enterprise.context.ApplicationScoped;
+
+import org.eclipse.microprofile.reactive.messaging.Message;
+import org.eclipse.microprofile.reactive.messaging.Outgoing;
+
+import com.ibm.ws.microprofile.reactive.messaging.fat.kafka.framework.AbstractDeliveryBean;
+
+/**
+ * Bean which drives the "input" channel
+ */
+@ApplicationScoped
+public class JsonbDeliveryBean extends AbstractDeliveryBean<TestData> {
+
+    @Override
+    @Outgoing("input")
+    public CompletionStage<Message<TestData>> getMessage() {
+        return super.getMessage();
+    }
+
+}

--- a/dev/com.ibm.ws.microprofile.reactive.messaging_fat/fat/src/com/ibm/ws/microprofile/reactive/messaging/fat/jsonb/JsonbReceptionBean.java
+++ b/dev/com.ibm.ws.microprofile.reactive.messaging_fat/fat/src/com/ibm/ws/microprofile/reactive/messaging/fat/jsonb/JsonbReceptionBean.java
@@ -1,0 +1,34 @@
+/*******************************************************************************
+ * Copyright (c) 2019 IBM Corporation and others.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ *
+ * Contributors:
+ *     IBM Corporation - initial API and implementation
+ *******************************************************************************/
+package com.ibm.ws.microprofile.reactive.messaging.fat.jsonb;
+
+import java.util.concurrent.CompletionStage;
+
+import javax.enterprise.context.ApplicationScoped;
+
+import org.eclipse.microprofile.reactive.messaging.Incoming;
+import org.eclipse.microprofile.reactive.messaging.Message;
+
+import com.ibm.ws.microprofile.reactive.messaging.fat.kafka.framework.AbstractReceptionBean;
+
+/**
+ * Bean for checking the "output" channel
+ */
+@ApplicationScoped
+public class JsonbReceptionBean extends AbstractReceptionBean<TestData> {
+
+    @Override
+    @Incoming("output")
+    public CompletionStage<Void> recieveMessage(Message<TestData> msg) {
+        return super.recieveMessage(msg);
+    }
+
+}

--- a/dev/com.ibm.ws.microprofile.reactive.messaging_fat/fat/src/com/ibm/ws/microprofile/reactive/messaging/fat/jsonb/JsonbServlet.java
+++ b/dev/com.ibm.ws.microprofile.reactive.messaging_fat/fat/src/com/ibm/ws/microprofile/reactive/messaging/fat/jsonb/JsonbServlet.java
@@ -1,0 +1,54 @@
+/*******************************************************************************
+ * Copyright (c) 2019 IBM Corporation and others.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ *
+ * Contributors:
+ *     IBM Corporation - initial API and implementation
+ *******************************************************************************/
+package com.ibm.ws.microprofile.reactive.messaging.fat.jsonb;
+
+import static org.junit.Assert.assertEquals;
+
+import java.time.Duration;
+import java.util.List;
+
+import javax.annotation.Resource;
+import javax.enterprise.concurrent.ManagedExecutorService;
+import javax.inject.Inject;
+import javax.servlet.annotation.WebServlet;
+
+import org.eclipse.microprofile.reactive.messaging.Message;
+import org.junit.Test;
+
+import componenttest.app.FATServlet;
+
+@WebServlet("/jsonbTest")
+public class JsonbServlet extends FATServlet {
+
+    private static final long serialVersionUID = 1L;
+
+    @Inject
+    private JsonbDeliveryBean deliveryBean;
+
+    @Inject
+    private JsonbReceptionBean receptionBean;
+
+    @Resource
+    private ManagedExecutorService executor;
+
+    @Test
+    public void testJsonb() throws InterruptedException {
+        TestData t = new TestData();
+        t.a = "test";
+        t.b = 42;
+        executor.submit(() -> deliveryBean.sendMessage(t));
+
+        List<Message<TestData>> receivedMessages = receptionBean.getReceivedMessages(1, Duration.ofSeconds(10));
+        TestData received = receivedMessages.get(0).getPayload();
+
+        assertEquals(t, received);
+    }
+}

--- a/dev/com.ibm.ws.microprofile.reactive.messaging_fat/fat/src/com/ibm/ws/microprofile/reactive/messaging/fat/jsonb/JsonbTest.java
+++ b/dev/com.ibm.ws.microprofile.reactive.messaging_fat/fat/src/com/ibm/ws/microprofile/reactive/messaging/fat/jsonb/JsonbTest.java
@@ -1,0 +1,55 @@
+/*******************************************************************************
+ * Copyright (c) 2019 IBM Corporation and others.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ *
+ * Contributors:
+ *     IBM Corporation - initial API and implementation
+ *******************************************************************************/
+package com.ibm.ws.microprofile.reactive.messaging.fat.jsonb;
+
+import org.jboss.shrinkwrap.api.ShrinkWrap;
+import org.jboss.shrinkwrap.api.spec.WebArchive;
+import org.junit.AfterClass;
+import org.junit.BeforeClass;
+import org.junit.runner.RunWith;
+
+import com.ibm.websphere.simplicity.ShrinkHelper;
+import com.ibm.websphere.simplicity.ShrinkHelper.DeployOptions;
+import com.ibm.ws.microprofile.reactive.messaging.fat.kafka.framework.AbstractDeliveryBean;
+import com.ibm.ws.microprofile.reactive.messaging.fat.kafka.framework.AbstractReceptionBean;
+
+import componenttest.annotation.Server;
+import componenttest.annotation.TestServlet;
+import componenttest.custom.junit.runner.FATRunner;
+import componenttest.topology.impl.LibertyServer;
+import componenttest.topology.utils.FATServletClient;
+
+@RunWith(FATRunner.class)
+public class JsonbTest extends FATServletClient {
+
+    public static final String APP_NAME = "jsob-messaging";
+    public static final String SERVER_NAME = "SimpleRxMessagingServer";
+
+    @Server(SERVER_NAME)
+    @TestServlet(servlet = JsonbServlet.class, contextRoot = APP_NAME)
+    public static LibertyServer server;
+
+    @BeforeClass
+    public static void setup() throws Exception {
+        WebArchive war = ShrinkWrap.create(WebArchive.class, APP_NAME + ".war")
+                        .addPackage(JsonbBean.class.getPackage())
+                        .addClasses(AbstractReceptionBean.class, AbstractDeliveryBean.class);
+
+        ShrinkHelper.exportDropinAppToServer(server, war, DeployOptions.SERVER_ONLY);
+        server.startServer();
+    }
+
+    @AfterClass
+    public static void cleanup() throws Exception {
+        server.stopServer();
+    }
+
+}

--- a/dev/com.ibm.ws.microprofile.reactive.messaging_fat/fat/src/com/ibm/ws/microprofile/reactive/messaging/fat/jsonb/TestData.java
+++ b/dev/com.ibm.ws.microprofile.reactive.messaging_fat/fat/src/com/ibm/ws/microprofile/reactive/messaging/fat/jsonb/TestData.java
@@ -1,0 +1,40 @@
+/*******************************************************************************
+ * Copyright (c) 2019 IBM Corporation and others.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ *
+ * Contributors:
+ *     IBM Corporation - initial API and implementation
+ *******************************************************************************/
+package com.ibm.ws.microprofile.reactive.messaging.fat.jsonb;
+
+import java.util.Objects;
+
+/**
+ * A very basic data class which can be serialized by jsonb
+ */
+public class TestData {
+
+    public String a;
+    public int b;
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(a, b);
+    }
+
+    @Override
+    public boolean equals(Object obj) {
+        if (this == obj)
+            return true;
+        if (obj == null)
+            return false;
+        if (getClass() != obj.getClass())
+            return false;
+        TestData other = (TestData) obj;
+        return Objects.equals(a, other.a) && b == other.b;
+    }
+
+}

--- a/dev/com.ibm.ws.microprofile.reactive.messaging_fat/fat/src/com/ibm/ws/microprofile/reactive/messaging/fat/kafka/ack/auto/KafkaAutoAckReceptionBean.java
+++ b/dev/com.ibm.ws.microprofile.reactive.messaging_fat/fat/src/com/ibm/ws/microprofile/reactive/messaging/fat/kafka/ack/auto/KafkaAutoAckReceptionBean.java
@@ -23,7 +23,7 @@ import org.eclipse.microprofile.reactive.messaging.Message;
 import com.ibm.ws.microprofile.reactive.messaging.fat.kafka.framework.AbstractReceptionBean;
 
 @ApplicationScoped
-public class KafkaAutoAckReceptionBean extends AbstractReceptionBean {
+public class KafkaAutoAckReceptionBean extends AbstractReceptionBean<String> {
 
     public static final String CHANNEL_IN = "auto-ack-in";
 

--- a/dev/com.ibm.ws.microprofile.reactive.messaging_fat/fat/src/com/ibm/ws/microprofile/reactive/messaging/fat/kafka/delivery/KafkaDeliveryBean.java
+++ b/dev/com.ibm.ws.microprofile.reactive.messaging_fat/fat/src/com/ibm/ws/microprofile/reactive/messaging/fat/kafka/delivery/KafkaDeliveryBean.java
@@ -20,7 +20,7 @@ import org.eclipse.microprofile.reactive.messaging.Outgoing;
 import com.ibm.ws.microprofile.reactive.messaging.fat.kafka.framework.AbstractDeliveryBean;
 
 @ApplicationScoped
-public class KafkaDeliveryBean extends AbstractDeliveryBean {
+public class KafkaDeliveryBean extends AbstractDeliveryBean<String> {
 
     public final static String CHANNEL_NAME = "delivery-test-output";
 

--- a/dev/com.ibm.ws.microprofile.reactive.messaging_fat/fat/src/com/ibm/ws/microprofile/reactive/messaging/fat/kafka/delivery/KafkaReceptionBean.java
+++ b/dev/com.ibm.ws.microprofile.reactive.messaging_fat/fat/src/com/ibm/ws/microprofile/reactive/messaging/fat/kafka/delivery/KafkaReceptionBean.java
@@ -22,7 +22,7 @@ import org.eclipse.microprofile.reactive.messaging.Message;
 import com.ibm.ws.microprofile.reactive.messaging.fat.kafka.framework.AbstractReceptionBean;
 
 @ApplicationScoped
-public class KafkaReceptionBean extends AbstractReceptionBean {
+public class KafkaReceptionBean extends AbstractReceptionBean<String> {
 
     public final static String CHANNEL_NAME = "reception-test-input";
 

--- a/dev/com.ibm.ws.microprofile.reactive.messaging_fat/fat/src/com/ibm/ws/microprofile/reactive/messaging/fat/kafka/framework/AbstractDeliveryBean.java
+++ b/dev/com.ibm.ws.microprofile.reactive.messaging_fat/fat/src/com/ibm/ws/microprofile/reactive/messaging/fat/kafka/framework/AbstractDeliveryBean.java
@@ -22,19 +22,19 @@ import org.eclipse.microprofile.reactive.messaging.Message;
  * <p>
  * To use this class, subclass it and override {@link #getMessage()} to add an {@code @Outgoing} annotation.
  */
-public class AbstractDeliveryBean {
+public class AbstractDeliveryBean<T> {
 
-    private final Queue<Message<String>> pendingMessages = new LinkedList<>();
-    private final Queue<CompletableFuture<Message<String>>> incompleteFutures = new LinkedList<>();
+    private final Queue<Message<T>> pendingMessages = new LinkedList<>();
+    private final Queue<CompletableFuture<Message<T>>> incompleteFutures = new LinkedList<>();
 
     public AbstractDeliveryBean() {
         super();
     }
 
-    public CompletionStage<Message<String>> getMessage() {
+    public CompletionStage<Message<T>> getMessage() {
         synchronized (this) {
             if (pendingMessages.isEmpty()) {
-                CompletableFuture<Message<String>> nextMessage = new CompletableFuture<>();
+                CompletableFuture<Message<T>> nextMessage = new CompletableFuture<>();
                 incompleteFutures.add(nextMessage);
                 System.out.println("Delivery bean returning incomplete CS");
                 return nextMessage;
@@ -53,14 +53,14 @@ public class AbstractDeliveryBean {
      * @param message the string to send in the message
      * @return a CompletableFuture which completes when the message is acknowledged
      */
-    public CompletableFuture<Void> sendMessage(String message) {
+    public CompletableFuture<Void> sendMessage(T message) {
         CompletableFuture<Void> ackCf = new CompletableFuture<>();
-        Message<String> msg = Message.of(message, () -> {
+        Message<T> msg = Message.of(message, () -> {
             ackCf.complete(null);
             return CompletableFuture.completedFuture(null);
         });
         synchronized (this) {
-            CompletableFuture<Message<String>> nextMessage = incompleteFutures.poll();
+            CompletableFuture<Message<T>> nextMessage = incompleteFutures.poll();
             if (nextMessage != null) {
                 System.out.println("Delivery bean passing message to incomplete CS");
                 nextMessage.complete(msg);

--- a/dev/com.ibm.ws.microprofile.reactive.messaging_fat/fat/src/com/ibm/ws/microprofile/reactive/messaging/fat/kafka/framework/AbstractReceptionBean.java
+++ b/dev/com.ibm.ws.microprofile.reactive.messaging_fat/fat/src/com/ibm/ws/microprofile/reactive/messaging/fat/kafka/framework/AbstractReceptionBean.java
@@ -28,11 +28,11 @@ import org.eclipse.microprofile.reactive.messaging.Message;
  * <p>
  * To use this class, subclass it and override {@link #recieveMessage(Message)} to add {@code @Incoming} and {@code @Acknowledgement} annotations.
  */
-public class AbstractReceptionBean {
+public class AbstractReceptionBean<T> {
 
-    private final Queue<Message<String>> receivedMessages = new LinkedList<>();
+    private final Queue<Message<T>> receivedMessages = new LinkedList<>();
 
-    public CompletionStage<Void> recieveMessage(Message<String> msg) {
+    public CompletionStage<Void> recieveMessage(Message<T> msg) {
         synchronized (this) {
             receivedMessages.add(msg);
             this.notifyAll();
@@ -40,8 +40,8 @@ public class AbstractReceptionBean {
         return CompletableFuture.completedFuture(null);
     }
 
-    public List<Message<String>> getReceivedMessages(int count, Duration timeout) throws InterruptedException {
-        ArrayList<Message<String>> result = new ArrayList<>();
+    public List<Message<T>> getReceivedMessages(int count, Duration timeout) throws InterruptedException {
+        ArrayList<Message<T>> result = new ArrayList<>();
         Duration remaining = timeout;
         long startTime = System.nanoTime();
         while (!remaining.isNegative() && result.size() < count) {
@@ -62,11 +62,11 @@ public class AbstractReceptionBean {
 
     /**
      * Return any messages which have been received by the bean, without waiting
-     * 
+     *
      * @return a list of messages, may be empty
      */
-    public List<Message<String>> getReceivedMessages() {
-        ArrayList<Message<String>> result = new ArrayList<>();
+    public List<Message<T>> getReceivedMessages() {
+        ArrayList<Message<T>> result = new ArrayList<>();
         synchronized (this) {
             while (!receivedMessages.isEmpty()) {
                 result.add(receivedMessages.poll());

--- a/dev/com.ibm.ws.microprofile.reactive.messaging_fat/fat/src/com/ibm/ws/microprofile/reactive/messaging/fat/kafka/partitions/PartitionTestReceptionBean.java
+++ b/dev/com.ibm.ws.microprofile.reactive.messaging_fat/fat/src/com/ibm/ws/microprofile/reactive/messaging/fat/kafka/partitions/PartitionTestReceptionBean.java
@@ -22,7 +22,7 @@ import org.eclipse.microprofile.reactive.messaging.Message;
 import com.ibm.ws.microprofile.reactive.messaging.fat.kafka.framework.AbstractReceptionBean;
 
 @ApplicationScoped
-public class PartitionTestReceptionBean extends AbstractReceptionBean {
+public class PartitionTestReceptionBean extends AbstractReceptionBean<String> {
 
     public static final String CHANNEL_NAME = "partition-test-incoming-topic";
 

--- a/dev/com.ibm.ws.microprofile.reactive.messaging_fat/fat/src/com/ibm/ws/microprofile/reactive/messaging/fat/suite/FATSuite.java
+++ b/dev/com.ibm.ws.microprofile.reactive.messaging_fat/fat/src/com/ibm/ws/microprofile/reactive/messaging/fat/suite/FATSuite.java
@@ -14,6 +14,7 @@ import org.junit.runner.RunWith;
 import org.junit.runners.Suite;
 import org.junit.runners.Suite.SuiteClasses;
 
+import com.ibm.ws.microprofile.reactive.messaging.fat.jsonb.JsonbTest;
 import com.ibm.ws.microprofile.reactive.messaging.fat.kafka.framework.tests.KafkaTestClientProviderTest;
 import com.ibm.ws.microprofile.reactive.messaging.fat.kafka.invalid.nolib.KafkaNoLibTest;
 import com.ibm.ws.microprofile.reactive.messaging.fat.loginModuleClassloading.LoginModuleClassloadingTest;
@@ -25,7 +26,8 @@ import com.ibm.ws.microprofile.reactive.messaging.fat.loginModuleClassloading.Lo
                 SaslPlainTests.class,
                 KafkaTestClientProviderTest.class,
                 LoginModuleClassloadingTest.class,
-                KafkaNoLibTest.class
+                KafkaNoLibTest.class,
+                JsonbTest.class
 })
 public class FATSuite {
 

--- a/dev/com.ibm.ws.microprofile.reactive.messaging_fat/publish/servers/SimpleRxMessagingServer/server.xml
+++ b/dev/com.ibm.ws.microprofile.reactive.messaging_fat/publish/servers/SimpleRxMessagingServer/server.xml
@@ -18,5 +18,7 @@
         <feature>localConnector-1.0</feature>
         <feature>mpReactiveMessaging-1.0</feature>
         <feature>servlet-4.0</feature>
+        <feature>jsonb-1.0</feature>
+        <feature>concurrent-1.0</feature>
     </featureManager>
 </server>

